libs/autotagging-commons/src/main/java/org/opensearch/rule/InMemoryRuleProcessingService.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.rule;
+
+import org.opensearch.autotagging.Attribute;
+import org.opensearch.autotagging.FeatureType;
+import org.opensearch.autotagging.Rule;
+import org.opensearch.rule.attribute_extractor.AttributeExtractor;
+import org.opensearch.rule.storage.AttributeValueStore;
+import org.opensearch.rule.storage.AttributeValueStoreFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * This class is responsible for managing in-memory view of Rules and Find matching Rule for the request
+ * Each auto-tagging feature should use a separate instance of this class as this avoid potential concurrency overhead
+ * in case of dynamic updates and attribute sharing scenarios
+ */
+public class InMemoryRuleProcessingService {
+
+    private final AttributeValueStoreFactory attributeValueStoreFactory;
+
+    /**
+     *  Constrcutor
+     * @param featureType
+     * @param attributeValueStoreSupplier
+     */
+    public InMemoryRuleProcessingService(
+        FeatureType featureType,
+        Supplier<AttributeValueStore<String, String>> attributeValueStoreSupplier
+    ) {
+        attributeValueStoreFactory = new AttributeValueStoreFactory(featureType, attributeValueStoreSupplier);
+    }
+
+    /**
+     * Adds the rule to in-memory view
+     * @param rule to be added
+     */
+    public synchronized void add(final Rule rule) {
+        new AddRuleOperation(rule, attributeValueStoreFactory).perform();
+    }
+
+    /**
+     * Removes the rule from in-memory view
+     * @param rule to be removed
+     */
+    public synchronized void remove(final Rule rule) {
+        new DeleteRuleOperation(rule, attributeValueStoreFactory).perform();
+    }
```

Comment:
Creating class for every operation seems bit of an overkill. Also synchronization is not really needed at this level, as every attributeStore can be modified concurrently. Although, I am assuming these operations are non-frequent

Reply:
>  Also synchronization is not really needed at this level, as every attributeStore can be modified concurrently. Although

Do you mean `can not be` ? Yes the frequency of these operations will be quite less frequent.

Reply:
I meant `AttributeStore` for different `AttributeType` can be modified concurrently. But right now, we are not even allowing that. Although given the infrequent scenarios, this should not be a problem

Reply:
Good Point! I think having more granular concurrency checks will be ideal to reduce concurrency overheads.